### PR TITLE
Refactor GELU and Sigmoid epilogue to use a common template (and add SiLu, Hardswish epilogue)

### PR DESCRIPTION
**Pull request by masahi**
Since Sigmoid and GeLU epilogue functors are almost identical, I propose to refactor them. The motivation is to make it easy to add another functors.

Now Sigmoid and GeLU epilogue functors are defined as:
```
using LinearCombinationSigmoid = LinearCombinationGeneric<Sigmoid, ElementOutput_, Count, ElementAccumulator_,
                                                          ElementCompute_, FloatRoundStyle::round_to_nearest>;

using LinearCombinationGELU = LinearCombinationGeneric<GELU, ElementOutput_, Count, ElementAccumulator_,
                                                          ElementCompute_, FloatRoundStyle::round_to_nearest>;
```

I'm going to add two new functors for `SiLU` and `HardSwish` activation, which can be implemented in the same manner.

What do you think? @hwu36

### CODE REVIEW

**Comment by hwu36** [comment]
@shangz-ai, is it you that told me that you need to use fp32 to compute sigmoid for the accuracy?

**Comment by masahi** [comment]
ok I've pushed two activation using this common template.

* SiLU https://pytorch.org/docs/stable/generated/torch.nn.SiLU.html
* HardSwish https://pytorch.org/docs/stable/generated/torch.nn.Hardswish.html

**Comment by masahi** [comment]
@d-k-b Thanks for the feedback, the doc is updated.

**Comment by hwu36** [comment]
Overall, looks good to me.  I think just change `FloatRoundStyle::round_to_nearest` to `Round` everywhere.

**Comment by masahi** [comment]
Oops thanks for spotting this, fixed.

**Comment by masahi** [comment]
Not sure why Ci has failed

File: include/cutlass/epilogue/thread/activation.h [hunk 1]
```diff
@@ -124,6 +124,61 @@ struct Sigmoid<Array<T, N> > {
...
+
+template <typename T, int N>
+struct HardSwish<Array<T, N> > {
+  CUTLASS_HOST_DEVICE
+  Array<T, N> operator()(Array<T, N> const &rhs) const {
+    Array<T, N> y;
+    HardSwish<T> hardswish_op;
+
+    CUTLASS_PRAGMA_UNROLL
+    for (int i = 0; i < int(rhs.size()); ++i) {
```
Comment (hwu36) [inline include/cutlass/epilogue/thread/activation.h:174]: Shall we replace `rhs.size()` to `N` to emphasize that it is a compile time constant?

**Comment by hwu36** [comment]
These 4 activations look expensive to me.  You can try to set `kIsHeavy` to `true`, check the generated code is different, and compare the performance to verify it.

If you setting them to `true`, the epilogue will not be fully unrolled which is good for the I$.  Usually complex activations such as `gelu` need to be set to `true`.

**Comment by masahi** [comment]
> These 4 activations look expensive to me. You can try to set `kIsHeavy` to `true`, check the generated code is different, and compare the performance to verify it.
> 
> If you setting them to `true`, the epilogue will not be fully unrolled which is good for the I$. Usually complex activations such as `gelu` need to be set to `true`.

My patch respects the current choice for the existing epilogues: `isHeavy = true` for GELU and whatever the default for Sigmoid which I assumed was `false` . For the two new activations, my reasoning was `false` for Silu because it is just sigmoid + mul, and also `false` for hardswish since it is similar to relu.

**Comment by hwu36** [comment]
SGTM.

`kIsHeavy` was newly introduced.  We benchmarked `gelu`, but not `sigmoid`.  If you find `sigmoid` is expensive in the future, you can create a PR and flip the bool.

**Comment by masahi** [comment]
Making sigmoid and silu `isHeavy = true` did improve performance on efficientnetv2 slightly: 8.58 msec -> 8.26 msec where 8.58 was the number from my PR https://github.com/apache/tvm/pull/9746. So I pushed that change. The change didn't help when the vectorized variant of sigmoid is used.

**Comment by hwu36** [comment]
> The change didn't help when the vectorized variant of sigmoid is used.

Vectorized one has less instruction and less burden to I$.

**Comment by hwu36** [comment]
> Making sigmoid and silu isHeavy = true did improve performance on efficientnetv2 slightly: 8.58 msec -> 8.26 msec where 8.58 was the number from my PR apache/tvm#9746.

You can update your TVM PR now. :)

**Review by hwu36 — APPROVED** [review]
